File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cs
using System;
using System.Collections.Generic;

public class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        int answer = 0;
        int sum = 0;
        
        Queue<int> q = new Queue<int>();
        
        for(int i = 0; i < bridge_length; i++)
        {
            q.Enqueue(0);
        }
        
        int idx = 0;
        
        while(idx < truck_weights.Length)
        {
            answer++;
            sum -= q.Dequeue();
            
            if(sum + truck_weights[idx] <= weight)
            {
                q.Enqueue(truck_weights[idx]);
                sum += truck_weights[idx];
                idx++;
            }
            else
            {
                q.Enqueue(0);
            }
        }
        answer += bridge_length;
        
        
        
        return answer;
    }
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.cpp
#include <string>
#include <vector>
#include <bits/stdc++.h>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    queue<int> bridge;
    
    int time = 0;
    int sum = 0;
    int idx = 0;
    
    for(int i = 0; i < bridge_length; i++)
    {
        bridge.push(0);
    }
    
    while(idx < truck_weights.size())
    {
        time++;
        
        sum -= bridge.front();
        bridge.pop();
        
        if(sum + truck_weights[idx] <= weight)
        {
            bridge.push(truck_weights[idx]);
            sum += truck_weights[idx];
            idx++;
        }
        else
        {
            bridge.push(0);
        }
    }
    time += bridge_length;
    
    return time;
}